controller/clustersecretupdater.go

```diff
@@ -0,0 +1,96 @@
+package controller
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/argoproj/argo-cd/controller/metrics"
+	appv1 "github.com/argoproj/argo-cd/pkg/apis/application/v1alpha1"
+	"github.com/argoproj/argo-cd/util/db"
+	"github.com/argoproj/gitops-engine/pkg/utils/kube/cache"
+	log "github.com/sirupsen/logrus"
+)
+
+type clusterSecretUpdater struct {
+	infoSource           metrics.HasClustersInfo
+	lock                 sync.Mutex
+	db                   db.ArgoDB
+	secretUpdateInterval time.Duration
+}
+
+func (c clusterSecretUpdater) Run(ctx context.Context) {
+	tick := time.Tick(c.secretUpdateInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			break
+		case <-tick:
+			for _, info := range c.infoSource.GetClustersInfo() {
+				cluster, err := c.db.GetCluster(context.Background(), info.Server)
```

Comment:
For unit testing I would suggest to extract code between lines 32~81 into separate function and test it independently . This way you don't need to have test that depends on time.